pkg/content/commit.go

```diff
@@ -0,0 +1,29 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package content
+
+import (
+	"github.com/GoogleContainerTools/kpt/pkg/content/extensions"
+	"github.com/GoogleContainerTools/kpt/pkg/location"
+)
+
+func Commit(content Content) (location.Location, error) {
+	switch content := content.(type) {
+	case extensions.ChangeCommitter:
+		return content.CommitChanges()
+	default:
+		return location.Location{}, nil
```

Comment:
That's an interesting one. The idea there is that if the content is a remote location type (like OCI) this would flush any changes made via filesys.FileSystem or kio.Writer. Returning nil was speculating what's the expectation if you flush changes to a location type like Dir (or a purely in-memory which is not implemented yet).

That is to say --- does nil a success because the changes have already been made to the dir destination? Or should a "not supported" err type be detected?

Returning nil was assuming if a destination arg (like in kpt fn sink) could be either oci:// directly or dir path, that every place that ended with a call to commit on when dest was a dir the "not supported" was "don't care - it's optional" then needing to test for not supported err is only extra work.

Reply:
Thoughts on that? It seems like a case where maybe a 3-valued return (location, bool, err) could be used --- with `{}, false, nil` being the not supported answer?

Reply:
some go packages use error value singular to signal special cases (io.EOF); that could work here too.

Reply:
Yeah, I think you're right. Similar to `return nil, ErrNotFound` instead of `return nil, nil` with the mental note "nil means not found"... Err matching "normal" conditions is always irritating, but there are places where `if err != nil { return err }` becomes `if err!=nil && !errors.Is(ErrNotFound) {return err}`

Maybe it's a lingering "dotnet-ism" --- there's a mantra "don't throw exceptions for non-exceptional circumstances" --- though that could just as much relate to the burden of forcing callers to add try/catch blocks.

Reply:
added content.ErrNotCommittable